lib/shelley/src/Cardano/Wallet/Shelley/Network/Blockfrost.hs

```diff
@@ -82,46 +100,60 @@ withNetworkLayer
     , syncProgress = undefined
     }
-
   where
+    currentNodeTip :: IO BlockHeader
+    currentNodeTip = runExceptT fetchLatestBlockHeader >>= \case
+        -- TODO: use cached value while retrying
+        Left err -> throwIO (BlockfrostException err)
+        Right header -> pure header
 
     watchNodeTip :: (BlockHeader -> IO ()) -> IO ()
     watchNodeTip callback = link =<< async (pollNodeTip callback)
+      where
+        pollNodeTip :: (BlockHeader -> IO ()) -> IO ()
+        pollNodeTip cb = forever $ do
+            runExceptT fetchLatestBlockHeader >>= \case
+                Left err -> throwIO (BlockfrostException err)
+                Right header ->
+                    bracketTracer (MsgWatcherUpdate header >$< tr) $ cb header
+            threadDelay 2_000_000
 
-    pollNodeTip :: (BlockHeader -> IO ()) -> IO ()
-    pollNodeTip callback = forever $ do
-        runExceptT fetchLatestBlockHeader >>= \case
-            Left err -> traceWith tr (MsgBlockfrostClientError err)
-            Right header ->
-                bracketTracer (MsgWatcherUpdate header >$< tr) $ callback header
-        threadDelay 2_000_000
+    fetchLatestBlockHeader :: ExceptT BlockfrostError IO BlockHeader
+    fetchLatestBlockHeader =
+        runBlockfrost BF.getLatestBlock >>= blockToBlockHeader
 
     runBlockfrost :: BF.BlockfrostClientT IO a -> ExceptT BlockfrostError IO a
     runBlockfrost =
         ExceptT . (first ClientError <$>) . BF.runBlockfrostClientT project
 
-    fetchLatestBlockHeader :: ExceptT BlockfrostError IO BlockHeader
-    fetchLatestBlockHeader = do
-        block@BF.Block{..} <- runBlockfrost BF.getLatestBlock
-        slotNo <- case _blockSlot of
-            Just (BF.Slot s) -> pure $ SlotNo $ fromIntegral s
-            Nothing -> throwError $ NoSlotError block
-        pure BlockHeader
-            { slotNo
-            , blockHeight = undefined -- Quantity "block" Word32
-            , headerHash = undefined -- !(Hash "BlockHeader")
-            , parentHeaderHash = undefined -- !(Maybe (Hash "BlockHeader"))
-            }
+blockToBlockHeader ::
+    forall m. MonadError BlockfrostError m => BF.Block -> m BlockHeader
+blockToBlockHeader block@BF.Block{..} = do
+    slotNo <- case _blockSlot of
+        Just s -> pure $ SlotNo $ fromIntegral $ BF.unSlot s
+        Nothing -> throwError $ NoSlotError block
+    blockHeight <- case _blockHeight of
+        Just height -> pure $ Quantity $ fromIntegral height
+        Nothing -> throwError $ NoBlockHeight block
+    headerHash <- parseBlockHeader _blockHash
+    parentHeaderHash <- for _blockPreviousBlock parseBlockHeader
+    pure BlockHeader { slotNo, blockHeight, headerHash, parentHeaderHash }
+  where
+    parseBlockHeader :: BF.BlockHash -> m (Hash "BlockHeader")
+    parseBlockHeader blockHash =
+        case fromText (BF.unBlockHash blockHash) of
+            Right hash -> pure hash
+            Left tde -> throwError $ InvalidBlockHash blockHash tde
```

Comment:
Would it make sense to use some helper function of type `:: Either e a -> ExceptT m e' a` here? 🤔

(That's the issue with explicit error handling: Large portions of the code become code that converts error types. 😅 )

Reply:
I used to use `hoist-error` library for these kind of tasks https://hackage.haskell.org/package/hoist-error-0.2.1.0
Ideally I'd introduce it to our codebase but its associated with additional effort of communicating and convincing team that its worth it and I don't feel like I am ready to kick-off such process 🤷🏻‍♂️

Reply:
Fair enough. 🤷🏻‍♂️ `hoist-error` may be a bit overkill, but I had the feeling that there could be a polymorphic function in `base` that does part of the job. E.g. `Data.Bifunctor.first` specializes to `first :: (e -> e') -> Either e a -> Either e' a`.

Reply:
@Unisay  @HeinrichApfelmus  well, we have
```
mapLeft]() :: (a -> c) -> Either a b -> Either c b
``` 
from https://hackage.haskell.org/package/either-5.0.1.1/docs/Data-Either-Combinators.html